### PR TITLE
Add some defensive asserts in MPolyQuo* constructors

### PR DESCRIPTION
**Pull request by thofma**

### CODE REVIEW

**Comment by thofma** [comment]
The errors seem to be coming from polymake. Is this related to the new binaries @benlorenz?

**Comment by benlorenz** [comment]
> The errors seem to be coming from polymake. Is this related to the new binaries @benlorenz?

Sorry about that, one new binary version will be yanked: https://github.com/JuliaRegistries/General/pull/44719

**Comment by thofma** [comment]
Thanks for looking into this.

**Comment by tthsqe12** [comment]
Are these === asserts meant to be permanent? Why sometimes asserts and other times error messages?

**Comment by thofma** [comment]
Yes, I want them permanent. The user facing ones should be errors. I will change them tomorrow.

File: src/Rings/MPolyQuo.jl [hunk 1]
```diff
@@ -648,16 +654,17 @@ lift(a::MPolyQuoElem) = a.f
...
-function (Q::MPolyQuo)(a::MPolyElem) 
-   if base_ring(Q) === parent(a)
-      return MPolyQuoElem(a, Q)
-   else
-      return MPolyQuoElem(Q.R(a), Q)
-   end
+function (Q::MPolyQuo{S})(a::S) where {S <: MPolyElem}
+   base_ring(Q) === parent(a) || error("Parent mismatch")
+   return MPolyQuoElem(a, Q)
+end
```
Comment (tthsqe12) [inline src/Rings/MPolyQuo.jl:664]: The juliaification of this `if else end` introduces a functional change. Now if the `a` has the right type (`S`) but the wrong parent, then we can't get coersion. This is my only gripe, and if this is intended, then everything is fine here.

Comment (thofma) [reply src/Rings/MPolyQuo.jl:664]: This is intended.

P.S.: I am actually dreaming of some `try_coerce(R, x)::Tuple{Bool, RingElement}` methods which whould get rid of many of these dirty hacks and at the same time allow useful error messages.

**Comment by fieker** [comment]
On Mon, Sep 13, 2021 at 03:21:09AM -0700, Tommy Hofmann wrote:
> @thofma commented on this pull request.
> 
> 
> 
> >     return a
>  end
>  
> -function (Q::MPolyQuo)(a::MPolyElem) 
> -   if base_ring(Q) === parent(a)
> -      return MPolyQuoElem(a, Q)
> -   else
> -      return MPolyQuoElem(Q.R(a), Q)
> -   end
> +function (Q::MPolyQuo{S})(a::S) where {S <: MPolyElem}
> +   base_ring(Q) === parent(a) || error("Parent mismatch")
> +   return MPolyQuoElem(a, Q)
> +end
> 
> This is intended.
> 
> P.S.: I am actually dreaming of some `try_coerce(R, x)::Tuple{Bool, RingElement}` methods which whould get rid of many of these dirty hacks and at the same time allow useful error messages.
> 
force_coerce(... throw_error = false)?

**Pull request merged**